File: src/main/java/org/example/service/CountryDataJpaService.java
package org.example.service;

import org.example.model.Country;
import org.example.persistence.data.jpa.CountryDataJpaRepository;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Optional;

@Service
@Transactional
public class CountryDataJpaService {

    private final CountryDataJpaRepository countryDataJpaRepository;

    public CountryDataJpaService(CountryDataJpaRepository countryDataJpaRepository) {
        this.countryDataJpaRepository = countryDataJpaRepository;
    }

    @Transactional(readOnly = true)
    public Optional<Country> findById(String code) {
        return countryDataJpaRepository.findById(code);
    }

    @Transactional(readOnly = true)
    public List<Country> findAll() {
        return countryDataJpaRepository.findAll();
    }

    public Country save(Country country) {
        return countryDataJpaRepository.save(country);
    }

}

File: src/main/kotlin/org/example/service/CountryDataJpaService.kt
package org.example.service

import org.example.model.Country
import org.example.persistence.data.jpa.CountryDataJpaRepository
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

/**
 * `open` not needed thanks to kotlin spring plugin.
 */
@Service
@Transactional
class CountryDataJpaService(private val countryDataJpaRepository: CountryDataJpaRepository) {

    @Transactional(readOnly = true)
    fun findById(code: String): Country? {
        return countryDataJpaRepository.findByIdOrNull(code)
    }

    @Transactional(readOnly = true)
    fun findAll(): List<Country> {
        return countryDataJpaRepository.findAll().toList()
    }

    fun save(country: Country): Country {
        return countryDataJpaRepository.save(country)
    }
}
